File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[327]区间和的个数.scala
package leetcode.editor.cn

object CountOfRangeSum {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    class Counter(nums: Array[Int], val lower: Int, val upper: Int) {
      var count = 0
      val preSum = Array.fill[Long](nums.length + 1)(0)
      val temp = Array.ofDim[Long](preSum.length)
      for i <- nums.indices do preSum(i + 1) = preSum(i) + nums(i)

      def getCount: Int =
        sort(0, preSum.length - 1)
        count

      def sort(low: Int, high: Int): Unit =
        if low < high then
          val mid = low + (high - low) / 2
          sort(low, mid)
          sort(mid + 1, high)
          merge(low, mid, high)

      def merge(low: Int, mid: Int, high: Int): Unit =
        for i <- low to high do temp(i) = preSum(i)

        var start = mid + 1
        var end = mid + 1
        for i <- low to mid do
          while start <= high && preSum(start) - preSum(i) < lower do start += 1

          while end <= high && preSum(end) - preSum(i) <= upper do end += 1

          count += end - start

        var i = low
        var j = mid + 1
        for p <- low to high do
          if j == high + 1 then
            preSum(p) = temp(i)
            i += 1
          else if i == mid + 1 then
            preSum(p) = temp(j)
            j += 1
          else if temp(i) > temp(j) then
            preSum(p) = temp(j)
            j += 1
          else
            preSum(p) = temp(i)
            i += 1
    }
    def countRangeSum(nums: Array[Int], lower: Int, upper: Int): Int =
      Counter(nums,lower,upper).getCount
  }
//leetcode submit region end(Prohibit modification and deletion)

  @main
  def runSum():Unit =
    Solution.countRangeSum(Array(-2,5,-1),-2,2)
}

File: cpp/leetcodecn/[327]区间和的个数.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Counter {
public:
    Counter(const vector<int> &nums, int lower, int upper) : preSum(nums.size() + 1, 0),
                                                             temp(nums.size() + 1, 0),
                                                             lower(lower), upper(upper) {
        for (auto i = 0; i < nums.size(); i++) {
            preSum[i + 1] = preSum[i] + nums[i];
        }
    }

    int getCount() {
        sort(0, preSum.size() - 1);
        return count;
    }

private:
    vector<int64_t> preSum;
    vector<int64_t> temp;
    int lower;
    int upper;
    int count = 0;

    void sort(int low, int high) {
        if (low >= high) return;

        auto mid = low + (high - low) / 2;
        sort(low, mid);
        sort(mid + 1, high);
        merge(low, mid, high);
    }

    void merge(int low, int mid, int high) {
        for (auto i = low; i <= high; ++i)
            temp[i] = preSum[i];

        auto start = mid + 1;
        auto end = mid + 1;
        for (auto k = low; k <= mid; ++k) {
            while (start <= high && preSum[start] - preSum[k] < lower)++start;
            while (end <= high && preSum[end] - preSum[k] <= upper) ++end;
            count += (end - start);
        }

        auto i = low;
        auto j = mid + 1;
        for (auto p = low; p <= high; p++) {
            if (j == high + 1) {
                preSum[p] = temp[i++];
            } else if (i == mid + 1) {
                preSum[p] = temp[j++];
            } else if (temp[i] > temp[j]) {
                preSum[p] = temp[j++];
            } else {
                preSum[p] = temp[i++];
            }
        }

    }
};

class Solution {
public:
    int countRangeSum(vector<int> &nums, int lower, int upper) {
        return Counter(nums, lower, upper).getCount();
    }
};
//leetcode submit region end(Prohibit modification and deletion)
#include <iostream>

int main() {
    vector<int> nums = {{-2, 5, -1}};
    cout << Solution().countRangeSum(nums, -2, 2);
}